File: MidtermLab3/MP Solver (Java)/MP_Solver.java
/**
 * Midterm Lab 3: CSV Dataset Processing
 * Student: LANCETA, ALEXANDER D.
 * Machine Problems: MP07, MP08, MP09
 */

import java.io.*;
import java.util.*;

public class MP_Solver {
    // Record class to structure dataset
    static class Record {
        String name, exam, result;
        int score;
        Record(String n, String e, String res, int s) {
            this.name = n; this.exam = e; this.result = res; this.score = s;
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter dataset file path: ");
        String path = sc.nextLine().replace("\"", "");
        List<Record> records = new ArrayList<>();

        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            String line;
            while ((line = br.readLine()) != null) {
                // SKIP metadata: Only process lines containing result keywords
                if (!line.contains("PASS") && !line.contains("FAIL")) continue;

                // Split commas ignoring those inside quotes
                String[] p = line.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
                
                if (p.length >= 8) {
                    String name = p[0].replace("\"", "").trim();
                    String exam = p[3].trim();
                    String result = p[7].trim();
                    int score = 0;
                    try { score = Integer.parseInt(p[6].trim()); } catch (Exception e) {}
                    records.add(new Record(name, exam, result, score));
                }
            }
        } catch (IOException e) {
            System.out.println("Error: " + e.getMessage());
        }

        // MP07: Sort and Print (Tabular Format)
        records.sort(Comparator.comparing(r -> r.exam));
        System.out.println("\n[MP07] Sorted by Exam Name:");
        for (int i = 0; i < records.size(); i++) {
            Record r = records.get(i);
            System.out.printf("%2d. %-45s | %s%n", (i + 1), r.exam, r.name);
        }

        // MP08: Filter
        System.out.println("\n[MP08] Candidates who passed:");
        for(Record r : records) {
            if("PASS".equalsIgnoreCase(r.result)) System.out.println("- " + r.name);
        }

        // MP09: Stats
        double avg = records.stream().mapToInt(r -> r.score).average().orElse(0.0);
        System.out.printf("\n[MP09] Dataset Statistics - Average Score: %.2f%n", avg);
        sc.close();
    }
}

File: MidtermLab3/MP Solver (Javascript)/MP_Solver.js
/**
 * Midterm Lab 3: CSV Dataset Processing
 * Student: LANCETA, ALEXANDER D.
 * Machine Problems: MP07, MP08, MP09
 */

const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({ 
    input: process.stdin, 
    output: process.stdout 
});

rl.question('Enter dataset file path: ', (path) => {
    try {
        const cleanPath = path.replace(/['"]/g, '');
        const fileContent = fs.readFileSync(cleanPath, 'utf8');
        
        const records = fileContent.split('\n')
            // CRITICAL: Filter for lines that actually contain PASS or FAIL.
            // This skips the first 6 metadata lines AND the header automatically!
            .filter(line => line.includes('PASS') || line.includes('FAIL'))
            .map(line => {
                // Splits on commas that are NOT inside double quotes
                const p = line.split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/);
                
                // Clean the data: remove quotes and whitespace
                const name = p[0] ? p[0].replace(/"/g, '').trim() : '';
                const exam = p[3] ? p[3].trim() : 'Unknown';
                const score = parseInt(p[6] ? p[6].trim() : '0');
                const result = p[7] ? p[7].trim() : '';
                
                return { name, exam, score, result };
            });

        // MP07: Sort by Exam Name
        records.sort((a, b) => a.exam.localeCompare(b.exam));
        console.log('\n[MP07] Sorted by Exam Name:');
        records.forEach((r, i) => console.log(`${i + 1}. ${r.exam} | ${r.name}`));

        // MP08: Filter by Keyword "PASS"
        console.log('\n[MP08] Candidates who passed:');
        const passed = records.filter(r => r.result === 'PASS');
        passed.forEach(r => console.log(`- ${r.name}`));

        // MP09: Stats (Average Score)
        const total = records.reduce((sum, r) => sum + r.score, 0);
        const avg = total / records.length;
        console.log(`\n[MP09] Dataset Statistics - Average Score: ${avg.toFixed(2)}`);

    } catch (err) {
        console.error('Error processing file:', err.message);
    }
    rl.close();
});